packages/datadog-plugin-koa/src/index.js

```diff
@@ -58,26 +58,62 @@ function createWrapRegister (tracer, config) {
     return function registerWithTrace (path, methods, middleware, opts) {
       const route = register.apply(this, arguments)
 
-      if (Array.isArray(path) || !route || !Array.isArray(route.stack)) return route
+      if (!Array.isArray(path) && route && Array.isArray(route.stack)) {
+        wrapStack(route)
+      }
 
-      route.stack = route.stack.map(middleware => {
-        if (typeof middleware !== 'function') return middleware
+      return route
+    }
+  }
+}
 
-        return function (ctx, next) {
-          if (!ctx || !web.active(ctx.req)) return middleware.apply(this, arguments)
+function createWrapRoutes (tracer, config) {
+  return function wrapRoutes (routes) {
+    return function routesWithTrace () {
+      const dispatch = routes.apply(this, arguments)
+      const dispatchWithTrace = function (ctx, next) {
+        if (!ctx.router) {
+          let router
+
+          Object.defineProperty(ctx, 'router', {
+            set (value) {
+              router = value
+              router.stack.forEach(layer => {
+                wrapStack(layer)
+              })
+            },
+
+            get () {
+              return router
+            }
+          })
+        }
 
-          web.exitRoute(ctx.req)
-          web.enterRoute(ctx.req, route.path)
+        return dispatch.apply(this, arguments)
+      }
 
-          return wrapMiddleware(middleware).apply(this, arguments)
-        }
-      })
+      dispatchWithTrace.router = dispatch.router
 
-      return route
+      return dispatchWithTrace
     }
   }
 }
 
+function wrapStack (layer) {
+  layer.stack = layer.stack.map(middleware => {
+    if (typeof middleware !== 'function') return middleware
+
+    return function (ctx, next) {
+      if (!ctx || !web.active(ctx.req)) return middleware.apply(this, arguments)
+
+      web.exitRoute(ctx.req)
+      web.enterRoute(ctx.req, layer.path)
+
+      return wrapMiddleware(middleware).apply(this, arguments)
```

Comment:
Could this `wrapMiddleware()` be done outside of this function (i.e. around line 105)?